File: examples/go/apps/alarms_api/alarm_queue.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Alarms API Example — Go WASM actor
//
// Demonstrates the Cloudflare Durable Objects alarm() pattern: a RequestQueue
// actor that batches incoming requests and processes them 10 seconds after the
// first write, using a durable alarm that survives actor deactivation.
//
// ## Cloudflare DO vs PlexSpaces Go
//
// | Cloudflare DO                             | PlexSpaces Go                          |
// |-------------------------------------------|----------------------------------------|
// | export class RequestQueue extends DO      | RequestQueue struct + BaseActor        |
// | this.ctx.storage.get('count')             | host.KV().Get("count")                 |
// | this.ctx.storage.put('count', n)          | host.KV().Put("count", val)            |
// | this.ctx.storage.setAlarm(Date.now()+10s) | host.Alarm().Set(nowMs + 10_000)       |
// | this.ctx.storage.getAlarm()               | host.Alarm().Get()                     |
// | async alarm() { ... }                     | case "__alarm__":                      |
// | new Response(JSON.stringify(result))      | return marshal(map[string]any{...})    |
// | wrangler.toml [[durable_objects]]         | app-config.toml [[supervisor.children]] |

package main

import (
	"encoding/json"
	"fmt"
	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

// ============================================================================
// RequestQueue Actor (Durable Object alarm() equivalent)
// ============================================================================

// QueueItem represents a single enqueued request.
type QueueItem struct {
	ID         int         `json:"id"`
	Data       interface{} `json:"data"`
	EnqueuedAt uint64      `json:"enqueued_at"`
}

// RequestQueue batches incoming items and processes them when the durable alarm fires.
// Each instance is an isolated actor — like a Durable Object per queue.
type RequestQueue struct {
	plexspaces.BaseActor

	// Queued items pending batch processing
	Items []QueueItem `json:"items"`

	// Current queue depth (mirrors len(Items) but durable via KV)
	Count int `json:"count"`

	// Lifecycle counters
	TotalProcessed   int `json:"total_processed"`
	TotalAlarmFires  int `json:"total_alarm_fires"`
}

// host is the package-level host function accessor (created once, reused across calls).
var host = plexspaces.NewHost()

// NewRequestQueue creates a new RequestQueue actor instance.
func NewRequestQueue() plexspaces.Actor {
	q := &RequestQueue{
		Items: make([]QueueItem, 0),
	}
	q.SetSelf(q)
	return q
}

// Init is called by the framework before the first Handle. Stores actor identity.
func (q *RequestQueue) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	q.SetRuntimeMetadata(config.ActorID)
	host.Info(fmt.Sprintf("RequestQueue %s: initialized", q.ActorID()))
	return ""
}

// Handle dispatches incoming messages to the appropriate handler.
func (q *RequestQueue) Handle(fromActor, msgType, payloadJSON string) string {
	switch msgType {
	case "enqueue":
		return q.handleEnqueue(payloadJSON)
	case "status":
		return q.handleStatus()
	case "reset":
		return q.handleReset()
	case "__alarm__":
		return q.handleAlarm()
	default:
		return marshal(map[string]any{"error": "unknown operation: " + msgType})
	}
}

// handleEnqueue adds an item to the queue.
// Sets a durable alarm on the FIRST item (10 seconds from now).
// Equivalent to Cloudflare DO:
//
//	if (count === 0) { await this.ctx.storage.setAlarm(Date.now() + 10_000) }
func (q *RequestQueue) handleEnqueue(payloadJSON string) string {
	var req struct {
		Item interface{} `json:"item"`
	}
	if err := json.Unmarshal([]byte(payloadJSON), &req); err != nil {
		req.Item = payloadJSON
	}
	if req.Item == nil {
		req.Item = fmt.Sprintf("item-%d", q.Count+1)
	}

	wasEmpty := q.Count == 0
	item := QueueItem{
		ID:         q.Count + 1,
		Data:       req.Item,
		EnqueuedAt: host.NowMs(),
	}
	q.Items = append(q.Items, item)
	q.Count++

	alarmSet := false
	if wasEmpty {
		// First item: schedule alarm 10 seconds from now.
		// Equivalent to: await this.ctx.storage.setAlarm(Date.now() + 10_000)
		fireAt := host.NowMs() + 10_000
		if err := host.Alarm().Set(fireAt); err != nil {
			return marshal(map[string]any{"error": "alarm_set failed: " + err.Error()})
		}
		alarmSet = true
		host.Info(fmt.Sprintf("RequestQueue %s: first item, alarm set for 10s at ts=%d",
			q.ActorID(), fireAt))
	}

	return marshal(map[string]any{
		"status":    "ok",
		"queued":    q.Count,
		"item_id":   item.ID,
		"alarm_set": alarmSet,
	})
}

// handleStatus returns the current queue depth and next alarm timestamp.
// Equivalent to Cloudflare DO: this.ctx.storage.getAlarm()
func (q *RequestQueue) handleStatus() string {
	fireAt, err := host.Alarm().Get()
	errMsg := ""
	if err != nil {
		errMsg = err.Error()
	}
	return marshal(map[string]any{
		"status":            "ok",
		"count":             q.Count,
		"alarm_at":          fireAt,
		"alarm_set":         fireAt > 0,
		"total_processed":   q.TotalProcessed,
		"total_alarm_fires": q.TotalAlarmFires,
		"error":             errMsg,
	})
}

// handleReset clears the queue and cancels any pending alarm (for test repeatability).
func (q *RequestQueue) handleReset() string {
	q.Items = make([]QueueItem, 0)
	q.Count = 0
	_ = host.Alarm().Delete()
	host.Info(fmt.Sprintf("RequestQueue %s: queue reset", q.ActorID()))
	return marshal(map[string]any{
		"status": "ok",
		"reset":  true,
	})
}

// handleAlarm is invoked by the framework when the scheduled alarm fires.
// Equivalent to Cloudflare DO: async alarm() { ... }
// Processes all batched items and clears the queue.
func (q *RequestQueue) handleAlarm() string {
	processed := q.Count
	q.TotalAlarmFires++
	q.TotalProcessed += processed

	host.Info(fmt.Sprintf("RequestQueue %s: alarm fired, processing %d items",
		q.ActorID(), processed))

	for _, item := range q.Items {
		host.Info(fmt.Sprintf("RequestQueue %s: processing item %d: %v",
			q.ActorID(), item.ID, item.Data))
	}

	// Clear the queue
	q.Items = make([]QueueItem, 0)
	q.Count = 0

	return marshal(map[string]any{
		"status":            "ok",
		"processed":         processed,
		"total_processed":   q.TotalProcessed,
		"total_alarm_fires": q.TotalAlarmFires,
	})
}

// ============================================================================
// Helpers
// ============================================================================

func marshal(v any) string {
	data, err := json.Marshal(v)
	if err != nil {
		return `{"error":"marshal failed"}`
	}
	return string(data)
}

// ============================================================================
// Registration
// ============================================================================

// init registers actors before the host calls any exported functions.
func init() {
	router := plexspaces.NewActorRouter()
	router.Route("RequestQueueActor", NewRequestQueue)
	plexspaces.Register(router)
}

func main() {}

File: examples/python/apps/alarms_api/alarm_queue.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (C) 2025 PlexSpaces Contributors
#
# Alarms API Example — Python WASM actor
#
# Demonstrates the Cloudflare Durable Objects alarm() pattern: a RequestQueue
# actor that batches incoming requests and processes them 10 seconds after the
# first write, using a durable alarm that survives actor deactivation.
#
# ## Cloudflare DO vs PlexSpaces Python
#
# | Cloudflare DO                             | PlexSpaces Python                      |
# |-------------------------------------------|----------------------------------------|
# | export class RequestQueue extends DO      | @actor class RequestQueueActor         |
# | this.ctx.storage.get('count')             | self.count  (state field)              |
# | this.ctx.storage.put('count', n)          | self.count = n  (auto-persisted)       |
# | this.ctx.storage.setAlarm(Date.now()+10s) | host.alarm.set(host.now_ms() + 10_000) |
# | this.ctx.storage.getAlarm()               | host.alarm.get()                       |
# | async alarm() { ... }                     | @handler("__alarm__")                  |
# | new Response(JSON.stringify(result))      | return {"status": "ok", ...}           |
# | wrangler.toml [[durable_objects]]         | app-config.toml [[supervisor.children]] |

from __future__ import annotations

import json

from plexspaces import actor, handler, host, state


# ─── RequestQueueActor ─────────────────────────────────────────────────────────


@actor(facets=["virtual_actor", "reminder"])
class RequestQueueActor:
    """Batches incoming requests and processes them when the durable alarm fires.

    State:
        items:             list of enqueued items (auto-persisted via actor state).
        count:             current queue depth.
        total_processed:   lifetime count of processed items.
        total_alarm_fires: lifetime count of alarm firings.
    """

    items: list = state(default_factory=list)
    count: int = state(default=0)
    total_processed: int = state(default=0)
    total_alarm_fires: int = state(default=0)

    @handler("enqueue")
    def enqueue(self, item: object = None) -> dict:
        """Add an item to the queue.

        Sets a durable alarm 10 seconds from now on the FIRST item only.
        Equivalent to Cloudflare DO:
            if (count === 0) { this.ctx.storage.setAlarm(Date.now() + 10_000) }
        """
        was_empty = self.count == 0
        queue_item = {
            "id": self.count + 1,
            "data": item,
            "enqueued_at": host.now_ms(),
        }
        self.items = self.items + [queue_item]
        self.count += 1

        alarm_set = False
        if was_empty:
            # First item: schedule alarm 10 seconds from now.
            fire_at = host.now_ms() + 10_000
            host.alarm.set(fire_at)
            alarm_set = True
            host.info(
                f"RequestQueue: first item queued, alarm set to fire in 10s at {fire_at}"
            )

        return {
            "status": "ok",
            "queued": self.count,
            "item_id": queue_item["id"],
            "alarm_set": alarm_set,
        }

    @handler("status")
    def status(self) -> dict:
        """Return current queue depth and next alarm timestamp.

        Equivalent to Cloudflare DO: this.ctx.storage.getAlarm()
        """
        alarm_at = host.alarm.get()
        return {
            "status": "ok",
            "count": self.count,
            "alarm_at": alarm_at,
            "alarm_set": alarm_at > 0,
            "total_processed": self.total_processed,
            "total_alarm_fires": self.total_alarm_fires,
        }

    @handler("reset")
    def reset(self) -> dict:
        """Clear queue and cancel pending alarm (for test repeatability)."""
        self.items = []
        self.count = 0
        host.alarm.delete()
        host.info("RequestQueue: queue reset")
        return {"status": "ok", "reset": True}

    @handler("__alarm__")
    def on_alarm(self) -> dict:
        """Process all queued items when the alarm fires.

        Equivalent to Cloudflare DO: async alarm() { ... }
        Delivered by the PlexSpaces reminder facet as the "__alarm__" message type.
        """
        processed = self.count
        self.total_alarm_fires += 1
        self.total_processed += processed

        host.info(f"RequestQueue: alarm fired, processing {processed} queued items")
        for item in self.items:
            host.info(
                f"RequestQueue: processing item {item['id']}: {json.dumps(item['data'])}"
            )

        # Clear the queue after processing
        self.items = []
        self.count = 0

        return {
            "status": "ok",
            "processed": processed,
            "total_processed": self.total_processed,
            "total_alarm_fires": self.total_alarm_fires,
        }


__all__ = ["RequestQueueActor"]
